File: lab07/2.py
import numpy as np

# Функция для вычисления правой части уравнения
def f(x, y):
    return -y * np.cos(x) + 0.5 * np.sin(2*x)

# Точное решение для сравнения
def exact_solution(x):
    return np.sin(x) - 1 + np.exp(-np.sin(x))

# Начальные условия
x0, y0 = 0, 0
x_end = 1
h = 0.1
n = int((x_end - x0) / h)

# Массивы для хранения значений
x = np.linspace(x0, x_end, n+1)
y = np.zeros(n+1)
y[0] = y0

# Классический метод Эйлера
for i in range(n):
    y[i+1] = y[i] + h * f(x[i], y[i])

# Вычисление точного решения и погрешности
exact_y = exact_solution(x)
errors = np.abs(exact_y - y)

# Вывод результатов
for i in range(n+1):
    print(f"x = {x[i]:.1f}, Numerical y = {y[i]:.6f}, Exact y = {exact_y[i]:.6f}, Error = {errors[i]:.6f}")

File: lab07/2.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

// Функция для вычисления правой части уравнения
double f(double x, double y) {
    return -y * cos(x) + 0.5 * sin(2*x);
}

// Точное решение для сравнения
double exact_solution(double x) {
    return sin(x) - 1 + exp(-sin(x));
}

int main() {
    // Начальные условия
    double x0 = 0.0, y0 = 0.0;
    double x_end = 1.0;
    double h = 0.1;
    int n = (int)((x_end - x0) / h);

    // Массивы для хранения значений
    double x[n+1], y[n+1];
    x[0] = x0;
    y[0] = y0;

    // Классический метод Эйлера
    for (int i = 0; i < n; ++i) {
        x[i+1] = x[i] + h;
        y[i+1] = y[i] + h * f(x[i], y[i]);
    }

    // Вывод результатов
    cout << fixed << setprecision(6);
    for (int i = 0; i <= n; ++i) {
        double exact_y = exact_solution(x[i]);
        double error = fabs(exact_y - y[i]);
        cout << "x = " << x[i] << ", Numerical y = " << y[i] << ", Exact y = " << exact_y << ", Error = " << error << endl;
    }

    return 0;
}
